src/Compilers/Core/Portable/Syntax/GreenNode.cs

```diff
@@ -904,34 +904,61 @@ private static bool EquivalentToInternal(GreenNode node1, GreenNode node2)
         public abstract SyntaxToken CreateSeparator<TNode>(SyntaxNode element) where TNode : SyntaxNode;
         public abstract bool IsTriviaWithEndOfLine(); // trivia node has end of line
 
-        public static GreenNode? CreateList(IEnumerable<GreenNode>? nodes, bool alwaysCreateListNode = false)
+        /*
+         * There are 3 overloads of this, because most callers already know what they have is a List<T> and only transform it.
+         * In those cases List<TFrom> performs much better.
+         * In other cases, the type is unknown / is IEnumerable<T>, where we try to find the best match.
+         * There is another overload for IROList, since most collections already implement this, so checking for it will
+         * perform better then copying to a List<T>, though not as good as List<T> directly.
+         */
+        public static GreenNode? CreateList<TFrom>(IEnumerable<TFrom>? enumerable, Func<TFrom, GreenNode> select)
+            => enumerable switch
+            {
+                null => null,
+                List<TFrom> l => CreateList(l, select),
+                IReadOnlyList<TFrom> l => CreateList(l, select),
+                _ => CreateList(enumerable.ToList(), select)
+            };
+
+        public static GreenNode? CreateList<TFrom>(List<TFrom> list, Func<TFrom, GreenNode> select)
         {
-            if (nodes == null)
+            switch (list.Count)
             {
-                return null;
+                case 0:
+                    return null;
+                case 1:
+                    return select(list[0]);
+                case 2:
+                    return SyntaxList.List(select(list[0]), select(list[1]));
+                case 3:
+                    return SyntaxList.List(select(list[0]), select(list[1]), select(list[2]));
+                default: {
+                    var array = new ArrayElement<GreenNode>[list.Count];
+                    for (int i = 0; i < array.Length; i++)
+                        array[i].Value = select(list[i]);
+                    return SyntaxList.List(array);
+                }
             }
+        }
 
-            var list = nodes.ToArray();
-
-            switch (list.Length)
+        public static GreenNode? CreateList<TFrom>(IReadOnlyList<TFrom> list, Func<TFrom, GreenNode> select)
```

Comment:
💭 This does make me wonder if we could somehow change this to only have one overload that takes `ReadOnlySpan<TFrom>`. Perhaps for the future...

Reply:
I also thought so. Note however that the only case where this is used is from WithLeading/Trailing trivia, and in those cases it seems more beneficial to check for IROList. See the List/Array benchmark above, it's not Span, but it indicates less special = better. And if I'm not missing anything ROSpan is a rare thing to be provided. (we could look into special casing for Array/Memory/ROMemory/Span/ROSpan and call into the ROSpan overload in all those cases. Will try and maybe open another PR)

Reply:
> And if I'm not missing anything ROSpan is a rare thing to be provided

`ReadOnlySpan<T>` cannot be cast to `IEnumerable<T>`, so to use this form all the callers would need to be updated to provide the correct type.